### PR TITLE
feat(multiple-auth): improvements in multiple auth implementation

### LINKED ISSUE #23: Improve multiple authentication support
We require the following enhancements in the multiple authentication feature:

- Throw validation exception named 'AuthValidationException' if any required auth parameter is missing or undefined
- Verify the 'OR' auth group implementation, we need to apply only the first available authentication scheme for 'OR' auth groups instead of selecting all of the auth schemes
- Add unit tests to cover the AuthValidationException and new logical changes to the apply_auth function

### PR DESCRIPTION
**Pull request by sufyankhanrao**
## What
This PR adds improvements in the implementation of multiple authentication schemes. The improvements include the renaming of auth-validation exception type and early returning in case of OR authentication group.

## Why
1. To fix a logical problem with the OR authentcation group
2. To make the custom exception name consistent across all languages

closes #23 

## Type of change
Select multiple if applicable.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause a breaking change)
- [ ] Tests (adds or updates tests)
- [ ] Documentation (adds or updates documentation)
- [ ] Refactor (style improvements, performance improvements, code refactoring)
- [ ] Revert (reverts a commit)
- [ ] CI/Build (adds or updates a script, change in external dependencies)

## Dependency Change
If a new dependency is being added, please ensure that it adheres to the following guideline https://github.com/apimatic/apimatic-codegen/wiki/Policy-of-adding-new-dependencies-in-the-core-libraries

## Breaking change
If the PR is introducing a breaking change, please ensure that it adheres to the following guideline https://github.com/apimatic/apimatic-codegen/wiki/Guidelines-for-maintaining-core-libraries

## Testing
List the steps that were taken to test the changes

## Checklist
- [x] My code follows the coding conventions
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have added new unit tests

### CODE REVIEW

File: test/test-apimatic-core/union_type_tests/any_of_test.rb [hunk 1]
```diff
@@ -248,10 +248,9 @@ def test_deserialize_nil_morning_type_any_of
     deserialized_morning = ApiHelper.json_deserialize(_morning_json)
     _any_of = _any_of.validate(deserialized_morning)
     actual_morning = _any_of.deserialize(nil)
-    expected_morning = nil
 
     assert _any_of.is_valid
-    assert_equal(expected_morning, actual_morning, 'Actual did not match the expected.')
+    assert_nil actual_morning
```
Comment (MaryamAdnan3) [inline test/test-apimatic-core/union_type_tests/any_of_test.rb:253]: why we are making changes in any-of tests area? Is it refactoring or will it cause any behavioral changes

File: test/test-apimatic-core/union_type_tests/one_of_test.rb [hunk 2]
```diff
@@ -133,10 +133,9 @@ def test_deserialize_nil_morning_type_one_of
     deserialized_morning = ApiHelper.json_deserialize(_morning_json)
     _one_of = _one_of.validate(deserialized_morning)
     actual_morning = _one_of.deserialize(nil)
-    expected_morning = nil
 
     assert _one_of.is_valid
-    assert_equal(expected_morning, actual_morning, 'Actual did not match the expected.')
+    assert_nil actual_morning
```
Comment (MaryamAdnan3) [inline test/test-apimatic-core/union_type_tests/one_of_test.rb:138]: same thing mentioned for anyof tests is applicable here

**Comment by MaryamAdnan3** [comment]
I have reviewed the changes for multiple auth or scheme early return and its impact on the or scheme test that you have refactored. I have also reviewed the exception name change and the changes look good to me

File: test/test-apimatic-core/union_type_tests/any_of_test.rb (hunk 1 above)
Comment (sufyankhanrao) [reply test/test-apimatic-core/union_type_tests/any_of_test.rb:253]: It is an enhancement that we can do against the testing framework that we are using currently which is minitest.

File: test/test-apimatic-core/union_type_tests/one_of_test.rb (hunk 2 above)
Comment (sufyankhanrao) [reply test/test-apimatic-core/union_type_tests/one_of_test.rb:138]: It is a refactoring that we can do against the testing framework that we are using currently which is minitest.

**Review by MaryamAdnan3 — APPROVED** [review]
LGTM

**Pull request merged**